src/module/iotjs_module_adc.c

```diff
@@ -152,107 +150,137 @@ void iotjs_adc_after_work(uv_work_t* work_req, int status) {
   }
 
   const iotjs_jval_t* jcallback = iotjs_adc_reqwrap_jcallback(req_wrap);
-  const iotjs_jval_t* jadc = iotjs_adc_get_jadc();
-  iotjs_make_callback(jcallback, jadc, &jargs);
+  iotjs_make_callback(jcallback, iotjs_jval_get_undefined(), &jargs);
 
   iotjs_jargs_destroy(&jargs);
 
   iotjs_adc_reqwrap_dispatched(req_wrap);
 }
 
 
-#define ADC_ASYNC(op)                                                        \
-  do {                                                                       \
-    uv_loop_t* loop = iotjs_environment_loop(iotjs_environment_get());       \
-    uv_work_t* req = iotjs_adc_reqwrap_req(req_wrap);                        \
-    uv_queue_work(loop, req, iotjs_adc_##op##_worker, iotjs_adc_after_work); \
-  } while (0)
+static void iotjs_adc_set_configuration(iotjs_adc_t* adc,
+                                        const iotjs_jval_t* jconfiguration) {
+#if defined(__linux__)
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_adc_t, adc);
+  iotjs_jval_t jdevice = iotjs_jval_get_property(jconfiguration, "device");
+  _this->device = iotjs_jval_as_string(&jdevice);
+  iotjs_jval_destroy(&jdevice);
+#elif defined(__NUTTX__)
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_adc_t, adc);
+  iotjs_jval_t jpin = iotjs_jval_get_property(jconfiguration, "pin");
+  _this->pin = iotjs_jval_as_number(&jpin);
+  iotjs_jval_destroy(&jpin);
+#endif
+}
 
 
-JHANDLER_FUNCTION(Export) {
-  JHANDLER_CHECK_ARGS(2, number, function);
+static void iotjs_adc_read_worker(uv_work_t* work_req) {
+  ADC_WORKER_INIT;
+  int32_t value = iotjs_adc_read(adc);
 
-  const iotjs_jval_t* jcallback = JHANDLER_GET_ARG(1, function);
+  if (value < 0) {
+    req_data->result = false;
+    return;
+  }
 
-  iotjs_adc_reqwrap_t* req_wrap =
-      iotjs_adc_reqwrap_create(jcallback, kAdcOpExport);
+  req_data->value = value;
+  req_data->result = true;
```

Comment:
In my opinion, it is better to handle error first with an if statement.
In this code, if `value` is less than 0, it is exceptions.